fix(datepicker): add footer with close button for screen readers

nigel5 opened the pull request:
Users with screen readers have difficulty closing the datepicker popup
obviously because they would have a hard time tapping the backdrop.
Also, on mobile device screen readers they do not have an escape key.

![datepickera11yclosedemo](https://user-images.githubusercontent.com/28766663/49682568-64e29c80-fa84-11e8-9fa4-e12af6e4888b.gif)

nigel5 commented:
issue #14429

crisbeto commented on src/lib/datepicker/calendar-footer.html:
```diff
@@ -0,0 +1,9 @@
+    <!-- The close button is for screen readers (it is invisible) -->
+    <button mat-button type="button" class="mat-calendar-close-button"
+            (click)="closeClicked()" [attr.aria-label]="closeCalendarLabel"
+            cdkAriaLive="polite">{{closeCalendarLabel}}
```
The element doesn't need `cdkAriaLive` since screen readers will read it out when it receives focus.

crisbeto commented on src/lib/datepicker/calendar-footer.html:
```diff
@@ -0,0 +1,9 @@
+<div class="mat-calendar-footer cdk-visually-hidden">
```
@jelbourn shouldn't we show the button if it receives focus? Otherwise people that are tabbing through the dialog without screen readers will land on it with no feedback. See how GitHub does it with their "Skip to content" button.

![github_-_google_chrome_2018-12-09_09-57-00](https://user-images.githubusercontent.com/4450522/49695317-95682a80-fb99-11e8-9ddf-32e492ed01a7.png)

crisbeto commented on src/lib/datepicker/calendar-footer.spec.ts:
```diff
@@ -0,0 +1,60 @@
+      ],
+      providers: [
+        MatDatepickerIntl,
+        {provide: Directionality, useFactory: () => ({value: 'ltr'})},
```
This isn't necessary, it'll default to `ltr` if a `Directionality` isn't provided.

crisbeto commented on src/lib/datepicker/calendar-footer.spec.ts:
```diff
@@ -0,0 +1,60 @@
+    TestBed.compileComponents();
+  }));
+
+  describe('standard calendar', () => {
```
It needs a test that the button does the correct thing when it's clicked.

crisbeto commented on src/lib/datepicker/calendar.ts:
```diff
@@ -174,9 +202,15 @@ export class MatCalendar<D> implements AfterContentInit, AfterViewChecked, OnDes
   @Input() headerComponent: ComponentType<any>;
 
+  /** An input indicating the type of the footer component, if set. */
+  @Input() footerComponent: ComponentType<any>;
```
We should have a test that checks whether this component is projected correctly.

nigel5 commented:
Hi @crisbeto ,

I appreciate your feedback. I have added the appropriate unit tests and removed the useless cdkAriaLive attribute from the button. Please keep me posted on whether the button should be shown when focused.

Thanks!

jelbourn commented:
@crisbeto you make a good point. We could make the button visible while focused by positioning it at the bottom-right corner as `position: absolute` on top of the calendar content.

nigel5 commented:
@jelbourn @crisbeto I updated the button so it shows when keyboard focused. this is what it looks like right now,

![datepickera11ydemo](https://user-images.githubusercontent.com/28766663/49980101-0ef87380-ff20-11e8-9b1c-1218a31ccd94.gif)

jelbourn commented:
@mmalerba can you take a look?

mmalerba commented on src/lib/datepicker/datepicker.ts:
```diff
@@ -145,6 +145,9 @@ export class MatDatepicker<D> implements OnDestroy, CanColor {
   @Input() calendarHeaderComponent: ComponentType<any>;
 
+  /** An input indicating the type of the custom footer component for the calendar, if set. */
+  @Input() calendarFooterComponent: ComponentType<any>;
```
I don't know about this, I don't think we want to let users drop content in here. At least not the way it's currently set up in this PR, the content will just wind up floating below the calendar.

For now how about just putting a button in `mat-datepicker-content` and we can consider opening it up via an API at some point in the future

nigel5 replied:
Sure

Orodan commented:
Is there any help needed to finish up this PR ? From what I can see the main content is ok and approved and the last work is about making the changes pass the CI right ?

MohammadRafik commented:
issue #14379 has been fixed by PR #20666
